bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs

```diff
@@ -130,6 +131,44 @@ public async Task DeleteManyByIdAsync(IEnumerable<Guid> ids)
         return policy == null ? (false, false) : (policy.Read, policy.Write);
     }
 
+    public async Task<IEnumerable<ServiceAccountSecretsDetails>> GetManyByOrganizationIdWithSecretsDetailsAsync(
+    Guid organizationId, Guid userId, AccessClientType accessType)
+    {
+        using var scope = ServiceScopeFactory.CreateScope();
+        var dbContext = GetDatabaseContext(scope);
+        var query = from sa in dbContext.ServiceAccount
+                    join ap in dbContext.ServiceAccountProjectAccessPolicy
+                        on sa.Id equals ap.ServiceAccountId into grouping
+                    from ap in grouping.DefaultIfEmpty()
+                    where sa.OrganizationId == organizationId
+                    select new
+                    {
+                        ServiceAccount = sa,
+                        AccessToSecrets = ap.GrantedProject.Secrets.Count(s => s.DeletedDate == null)
+                    };
+
+        query = accessType switch
+        {
+            AccessClientType.NoAccessCheck => query,
+            AccessClientType.User => query.Where(c =>
+                c.ServiceAccount.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Read) ||
+                c.ServiceAccount.GroupAccessPolicies.Any(ap =>
+                    ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Read))),
+            _ => throw new ArgumentOutOfRangeException(nameof(accessType), accessType, null),
+        };
+
+        var results = (await query.ToListAsync())
+            .GroupBy(g => g.ServiceAccount)
+            .Select(g =>
+                new ServiceAccountSecretsDetails
+                {
+                    ServiceAccount = Mapper.Map<Core.SecretsManager.Entities.ServiceAccount>(g.Key),
+                    AccessToSecrets = g.Sum(x => x.AccessToSecrets),
+                }).OrderBy(c => c.ServiceAccount.RevisionDate).ToList();
```

Comment:
This will generate a SQL query similar to 

```
  SELECT [s].[Id], [s].[CreationDate], [s].[Name], [s].[OrganizationId], [s].[RevisionDate], (
          SELECT COUNT(*)
          FROM [ProjectSecret] AS [p0]
          INNER JOIN [Secret] AS [s0] ON [p0].[SecretsId] = [s0].[Id]
          WHERE (([p].[Id] IS NOT NULL) AND ([p].[Id] = [p0].[ProjectsId])) AND ([s0].[DeletedDate] IS NULL)) AS [AccessToSecrets]
      FROM [ServiceAccount] AS [s]
      LEFT JOIN (
          SELECT [a].[GrantedProjectId], [a].[ServiceAccountId]
          FROM [AccessPolicy] AS [a]
          WHERE [a].[Discriminator] = N'service_account_project'
      ) AS [t] ON [s].[Id] = [t].[ServiceAccountId]
      LEFT JOIN [Project] AS [p] ON [t].[GrantedProjectId] = [p].[Id]
      WHERE [s].[OrganizationId] = <org_id_here>
```

This will fetch a list of service accounts with a `AccessToSecrets` column with the number of secrets per project the service account has permission to. 


Then at the .NET layer, the calculations for grouping by service accounts and summing the total access to secrets are preformed. 

I tried several patterns to try to have everything performed in the SQL query but seems like EF has some limitations with [group by](https://learn.microsoft.com/en-us/ef/core/querying/complex-query-operators#groupby).

Open to feedback if anyone has a better approach.

Reply:
I think this looks great 👍